File: Test/FrameAnimation.cpp
#include "Application.h"
#include "Callback.h"
#include "FrameAnimPanel.h"
#include "Image.h"
#include "Label.h"
#include "MainWindow.h"

using namespace d14uikit;

#define DEMO_NAME L"FrameAnimation"

int main(int argc, char* argv[])
{
    float dpi = 96.0f;
    if (argc >= 2 && strcmp(argv[1], "HighDPI") == 0)
    {
        dpi = 192.0f;
    }
    Application app(DEMO_NAME, dpi);
    app.setSize({ 300, 336 });

    //------------------------------------------- Initialize UI objects.

    MainWindow mwnd(DEMO_NAME);
    mwnd.setMinimizeButton(false);
    mwnd.setMaximizeButton(false);

    Panel clntArea;
    mwnd.setContent(&clntArea);

    FrameAnimPanel animArea;
    animArea.setParent(&clntArea);
    animArea.setPosition({ 22, 22 });
    animArea.setSize({ 256, 256 });

    std::vector<Image*> framePtrs;
    std::vector<std::unique_ptr<Image>> frames;
    for (int i = 0; i < 12; ++i)
    {
        auto frame = std::make_unique<Image>(L"images/stick_boy/" + std::to_wstring(i) + L".png");

        framePtrs.push_back(frame.get());
        frames.push_back(std::move(frame));
    }
    animArea.setFrames(framePtrs);
    // Applies point-sampler to make the pixel art clear.
    animArea.setFrameInterpMode(BitmapInterpMode::NearestNeighbor);
    animArea.setFrameTimeSpan(0.06f);

    // Notifies the renderer to draw frames continuously.
    animArea.setAnimState(true);

    Label fpsLbl(L"FPS: None");
    fpsLbl.setParent(&clntArea);
    fpsLbl.setPosition({ 22, 22 });

    //------------------------------------------- Set UI event callbacks.

    // This will be called before the renderer drawing each frame.
    fpsLbl.D14_onUpdate(p, &)
    {
        static int fps = 0;
        if (app.fps() != fps)
        {
            fps = app.fps();
            ((Label*)p)->setText(L"FPS: " + std::to_wstring(fps));
        }
    };
    return app.run();
}

File: Test/PyBind/FrameAnimation.py
from sys import argv

from D14UIKit import *

DEMO_NAME = 'FrameAnimation'

if __name__ == '__main__':
    dpi = 96.0
    if len(argv) >= 2 and argv[1] == 'HighDPI':
        dpi = 192.0

    app = Application(DEMO_NAME, dpi)
    app.size = Size(300, 336)

    #------------------------------------------- Initialize UI objects.

    mwnd = MainWindow(DEMO_NAME)
    mwnd.minimizeButton = False
    mwnd.maximizeButton = False

    clntArea = Panel()
    mwnd.content = clntArea

    animArea = FrameAnimPanel()
    animArea.parent = clntArea
    animArea.position = Point(22, 22)
    animArea.size = Size(256, 256)

    animArea.frames = [Image(f'images/stick_boy/{i}.png') for i in range(12)]
    # Applies point-sampler to make the pixel art clear.
    animArea.frameInterpMode = BitmapInterpMode.NearestNeighbor
    animArea.frameTimeSpan = 0.06

    # Notifies the renderer to draw frames continuously.
    animArea.animState = True

    fpsLbl = Label('FPS: None')
    fpsLbl.parent = clntArea
    fpsLbl.position = Point(22, 22)

    #------------------------------------------- Set UI event callbacks.

    # This will be called before the renderer drawing each frame.
    def displayFPS(p):
        if app.fps != displayFPS.fps:
            displayFPS.fps = app.fps
            p.text = f'FPS: {displayFPS.fps}'

    displayFPS.fps = 0
    fpsLbl.f_onUpdate = displayFPS

    exit(app.run())
